### PR TITLE
Implement printf

### PR DESCRIPTION
**Pull request by callumfare**
Clspv provides a printf definition that stores the value of each argument in a program-wide printf storage buffer. This implementation uses the appropriate ClspvReflection instructions to parse this data and print the resulting strings at runtime.

This fixes all failing printf CTS tests, other than one which requires physical addressing (to print the value of a pointer).

Related Pull Requests:
https://github.com/KhronosGroup/SPIRV-Registry/pull/168
https://github.com/KhronosGroup/SPIRV-Headers/pull/290
https://github.com/google/clspv/pull/925

This contribution is being made by Codeplay on behalf of Samsung.

### CODE REVIEW

File: src/program.hpp
Comment (kpet) [inline src/program.hpp:301]: This doesn't really need to be dynamically allocated, does it? It's not really a performance concern given the path it's happening on but I thought I'd ask in case I missed a good reason for it.

File: src/queue.cpp
Comment (kpet) [inline src/queue.cpp:799]: Any reason you're not using memset directly on a host-mapped pointer to the buffer's memory? The way it's done looks very inefficient. Taking a step back: do you really need to zero the entire buffer? With the default size, we're talking about 1MB's worth of zeroing for each kernel instance! Am I missing something?

**Review by kpet — CHANGES_REQUESTED** [review]
Many thanks for this contribution :). I'm starting to wrap my head around it and already have a few comments. I still need to do a full review but thought I'd share what I've found so far.

File: src/program.hpp
Comment (callumfare) [reply src/program.hpp]: It doesn't, I think my original thought here was that `m_printf_buffer_info` is optional so it may or may not be allocated for a program. I've updated this, now the buffer size defaults to 0 which is used to represent the absence of printf buffer info.

File: src/queue.cpp
Comment (callumfare) [reply src/queue.cpp]: No good reason for not using memset on the mapped memory directly, I've fixed that now. An older version did require the whole buffer to be zeroed, but that's not the case any more. I've updated this to only zero the first 4 bytes (i.e. the allocation counter).

File: src/kernel.cpp
Comment (callumfare) [reply src/kernel.cpp:228]: It definitely isn't optimal, I did it this way to closely match the existing implementation for program-scope constant data, since notionally the printf buffer is just another program-scope buffer (although not constant). This did get complicated by the possibility of different kernel instances writing to the same buffer, so it doesn't live in the `cvk_program`, but I'm not 100% sure that what I've done is always going to be correct.

Before I make any changes here, do you have any ideas about the best way to handle the buffer? I suspect it makes the most sense for there to be one per program per queue, although if there are multiple kernels in-flight then should the `cvk_executor_thread` actually own the buffer?

File: src/queue.cpp
Comment (kpet) [reply src/queue.cpp]: Looks good, resolving.

File: src/kernel.cpp
Comment (kpet) [reply src/kernel.cpp:228]: There's a spectrum of possible options, with different trade-offs. Here are two options with some of the pros and cons:
1. One buffer per kernel instance, i.e. what this PR currently implements. With this option we definitely don't have to worry about concurrent accesses from multiple kernel instances and only concurrency between the invocations of a single kernel instance has to be managed. However this requires allocating one buffer per kernel instance that uses printf which is rather costly. We could add a pool of printf buffers to amortise the cost of those allocations but then we'd have the problem of managing the recycling of the memory (unless we kept as many buffers as the peak number of kernel instances reached; with 1K kernels in flight and a default buffer size of 1MiB, we're already talking about 1GiB of memory). It is possible to restrict the allocation of buffers to those kernels instances that do use printf which would limit the cost but could still result in undesirable pathological cases (e.g. lots of kernel instances making very little use of printf).
2. One buffer per queue. The allocation troubles essentially vanish but we now have to manage concurrent accesses from multiple kernel instances. clvk only supports in-order queues for now but a single kernel instance can result in multiple Vulkan compute dispatches that do run concurrently. Having said that, I think this could be managed correctly with the same approach currently (see clspv PR) used to manage concurrency between the work-items of a kernel instance provided the atomics use a `Device` scope. Alternatively clvk could serialise compute dispatches that make use of the printf buffer. This would also simplify the management of the completion logic to format and print those strings in clvk (only one buffer has to be considered). Maybe we'd want to increase the default size of the buffer a bit, at least until OpenCL standardises additional control over it like in the cl_arm_printf extension for example. The buffer could even be created lazily the first time a kernel that makes use of printf is enqueued (if we do that I'd quite like to see a bitfield declaring features with runtime impact individual kernels use added to the non-semantic instruction set).

Assuming I haven't made a mistake in the above reasoning, I can't really think of a clear downside to going for 2 right now (though it's Sunday evening :-)). Can you think of other issues with this approach?

Comment (callumfare) [reply src/kernel.cpp]: The second option definitely makes the most sense to me, so I've made some changes based on that.

- The printf buffer is now in the queue
- The printf buffer is lazily created when the first kernel that may use printf is enqueued
- Additional pipeline barriers are used to serialize execution of different region dispatches for kernels that use printf
- Kernels that use printf are considered non-batchable. This prevents the case where multiple kernels from different programs try to use printf in the same batched command group, as they may both write to the printf buffer and interfere with each other. 
- The actually printing happens as soon as the command is finished.

I agree with your idea of adding a bitfield to the non-semantic instruction set to allow things to work on a per-kernel basis. Are there any other flags you think should be added other than a 'uses printf' one? I can add any others to my SPIRV-Headers change in the same PR. There are a couple of TODOs in the latest commits that will get resolved when I make those changes.

**Comment by callumfare** [comment]
I've updated this PR. As with the clspv change, it's based on top of the physical module constants change (#481) to make merging easier, so only the top commit is relevant here. It can be rebased when that change goes in. It also requires the SPIRV-Headers module bump from that PR.

**Comment by callumfare** [comment]
I've updated/rebased this. I believe everything on the clvk side is finished, but the matching clspv changes are currently blocked by a regression after the latest rebase. Once the clspv PR is sorted and merged this should be good to go.

**Comment by rjodinchr** [comment]
I am taking over this PR

**Comment by lpavank** [comment]
@rjodinchr, Gowtham from Samsung did make progress (which is being internally reviewed) and plan to upload what we got sometime this week. Is that OK or still want to take over?

**Comment by kpet** [comment]
Superseded by #565. @callumfare Many thanks for all the great work! You got us 99% there :).

**Pull request closed without merging**
